[REF] website, *: headers redesign

Brieuc-brd opened the pull request:
This PR redesigns all headers and improves and adapts some related features.

**New Header redesigns:** 
All previous header designs have been removed in favor of the new redesigned set of headers.

**New Elements feature:**
This new feature allows the user to select the elements he wants to display in the header (e.g. social links, call to action, ...)

**Mobile menu feature:**
Since we decided to keep the mobile version with an "off canvas" layout, the Mobile Menu feature has been removed.

**Alignment feature:**
This feature has been adapted to match the new headers.

**Mobile Alignment Feature:** 
Since the mobile header layout is always "off canvas", this feature allows the user to choose the menu alignment for the mobile version.

Requires: 
- https://github.com/odoo/design-themes/pull/651
- https://github.com/odoo/upgrade/pull/5214

task-3060986
task-3474743

Co-authored-by: Brieuc-brd [brd@odoo.com](mailto:brd@odoo.com)
Co-authored-by: robinlej [role@odoo.com](mailto:role@odoo.com)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/119650).

Brieuc-brd commented:
@stefanorigano 👋  here's the PR about the new headers.

stefanorigano commented on addons/website/static/src/scss/website.scss:
```diff
@@ -247,7 +247,16 @@ $-seen-urls: ();
+// The solution is to manipulate the selector according to the chosen header.
+$o_navbarlight_selector:'.navbar-light';
+@if o-website-value('header-template') == 'salesFour' {
+    $o_navbarlight_selector: '.navbar-light:not(.o_header_salesFour_top_part)';
```
Rather than creating an exception for a template's component specifically, I'd prefer to don't trigger these rules for navs that contain a generic class, eg `.navbar-light:not(.o-navbar-light-static)`.

Brieuc-brd replied:
That's what I did earlier, but as I wrote in the comment above these lines, changing this selector globally breaks the default `nav-link` style for the other templates. That's why I created this condition to change this selector only for the template that need it.
![Capture d’écran 2023-04-26 à 08 29 41](https://user-images.githubusercontent.com/80679690/234489346-3b32eae9-722f-47a4-b971-a75551da0862.png)

Brieuc-brd replied:
The code has been updated and this part is no longer necessary 👍

stefanorigano reviewed (changes requested):
@Brieuc-brd , very good thanks. Some minors about the code

Brieuc-brd commented:
- Fixed issues
- Camel case and `_part` removed
- Fixed an issue related to the "navigation style" feature

I'm waiting for the 🟢  state

stefanorigano commented on addons/website/views/website_templates.xml:
```diff
@@ -259,1105 +259,1166 @@
+    <t t-call="website.navbar">
+        <t t-set="_navbar_classes" t-valuef="o_header_mobile d-block d-lg-none shadow-sm"/>
+        <t t-set="_navbar_expand_class" t-valuef="None"/>
+        <div id="top_menu_container" class="container flex-wrap justify-content-between">
```
I think it's the right occasion to correctly prefix these IDs. What about 

```xml
<div id="wrapwrap">
    <header id="o_main_nav">
        <nav id="o_main_nav_top"/>
        <nav id="o_main_nav_bottom"/>
    </header>
</div>```

stefanorigano commented on addons/website/views/website_templates.xml:
```diff
@@ -259,1105 +259,1166 @@
+            <t t-call="website.placeholder_header_brand"/>
+            <ul class="list-unstyled d-flex align-items-center gap-2 mb-0">
+                <!-- Cart / Wishlist placeholder -->
+                <div id="o_cart_whish_placeholder_mobile"/>
```
edit: In any case better to don't reference the wishlist module in the placeholder name, let's keep it more generic ;)

Brieuc-brd replied:
> An `<ul>` can't contain a `<div>`.

@stefanorigano 
Since the Extra elements, Cart and Dropdown elements have different parent tags (`div` or `li`), and we sometimes need `div` instead of `li`, can I generalize all these elements with a `div` wrapper instead of `ul`/`li` ?

stefanorigano replied:
Sure,  a `<nav>` tag would be even more appropriate IMO.

Brieuc-brd replied:
Since the main wrap  tag is already a `<nav>` : 
```xml
<header>
<nav data-name="Navbar" class="navbar navbar-expand-lg ...">
...
</nav>
</header>

```
 I'm not sure it's a good idea to nest a '<nav>' inside another. What do you think? 🤔 
I think we should go with `div`

Brieuc-brd replied:
As the Portal default template uses `ul/li`, I've changed the header templates to match it and used `nav` when it was possible.

Brieuc-brd commented on addons/website/static/src/scss/website.scss:
```diff
@@ -257,10 +257,37 @@ $-seen-urls: ();
         @extend .navbar-dark;
+
+        .btn-close {
+            @extend .btn-close-white;
```
Indeed, I can't toggle classes for that.
I replaced the `@extend` by the raw scss used in this extension.

Brieuc-brd commented on addons/website/views/website_templates.xml:
```diff
@@ -259,1105 +259,1166 @@
+                <div id="o_cart_whish_placeholder_mobile"/>
+                <li>
+                    <button
+                        class="nav-link btn me-auto p-2 fs-5 o_not_editable"
```
In the end this was not necessary, I removed it

Brieuc-brd commented:
> Hello @Brieuc-brd , thanks 👍 Some minor remarks about CSS and XML.
> 
> Let me know when I can check the design ;)

Hello @stefanorigano 👋 

I've made the changes 👍 
The branch isn't green yet, but I'll let you check the design, then I'll rebase the branch and make it green.

stefanorigano reviewed (changes requested):
Hello @Brieuc-brd , some remarks for you in the code ;)
I have also some remarks concerning design and usability, I'll post everything here for simplicity.

(pde pinged on discord)

~~I think that the **border** option should affect the `border-bottom` property only. There are cases where you want a "complete border" but it's much less common IMO.~~
edit: @Brieuc-brd , as @pierre-pde said box-shadow can be used instead 👍 

---

The label should be place at the field's top
<img width="1063" alt="image" src="https://github.com/odoo/odoo/assets/8558351/3fbb20f4-b228-4922-b4d0-267f06546a13">

---

I find this dropdown-menu "displacement" very problematic. Scenario:
1 - Click to open the dropdown
2 - (re)clicking in the same pont will log me out (😬) since the toggler's place will be taken by the menu

https://github.com/odoo/odoo/assets/8558351/b8102ce5-197d-4126-ad56-a12d246c0f2e

I think we have 2 options:
1. We "unfold " the menu by default (I suggest this)
2. We add a `dropup` class 👇🏼.

https://github.com/odoo/odoo/assets/8558351/6cdcc3c7-6050-446d-a1ea-c7b71458b1e8

---

Searchbar's suggestions displace navigation links. Trying to hide them I clicked on big "X" closing the entire offcanvas menu instead .

https://github.com/odoo/odoo/assets/8558351/8b7c4b28-c3b0-466a-9eb0-b00d4a9cc19c

---

Not sure if already on master, but mega-menus are unfolded by default when inside the "+" dropdown

https://github.com/odoo/odoo/assets/8558351/3277546b-2fbc-4bf8-a6d0-aca381a6a675

---

For some reason "Contact us" becomes `btn-lg`on mobile.
![image](https://github.com/odoo/odoo/assets/8558351/508de27a-d77d-40e3-a407-0905da95fdd5)

---
**Stretch-menu**  template ( + **Vertical-menu** and others): "Over the content" -> Undesidered color changes


https://github.com/odoo/odoo/assets/8558351/0289f1d6-f703-4215-83a9-51d64b130619

---

User's dropdown-menu should be right-alighed
![image](https://github.com/odoo/odoo/assets/8558351/719c4987-3ad9-4924-8ffe-bbd0464c7760)

---

Search overlay modal: Border/overflow issue 
![image](https://github.com/odoo/odoo/assets/8558351/6e91f7ba-f6a7-491e-bb07-fb2866d9e4ef)

---

Social links (top-right). Aren't these supposed to become white?
![image](https://github.com/odoo/odoo/assets/8558351/a37f809e-94da-4071-8389-1a7f8ffe1ab5)

Brieuc-brd commented on addons/portal/views/portal_templates.xml:
```diff
@@ -51,21 +51,22 @@
                 <div t-attf-class="dropdown-menu #{_dropdown_menu_class}" role="menu">
                     <t t-foreach="languages" t-as="lg">
-                        <a t-att-href="url_for(request.httprequest.path + '?' + keep_query(), lang_code=lg[0])"
+                        <a class="dropdown-item" t-att-href="url_for(request.httprequest.path + '?' + keep_query(), lang_code=lg[0])"
```
I added it just below 👇 :) (https://github.com/odoo/odoo/pull/119650/files#diff-823e5db841dca1798ff1300e243059a4e1c93343598d2be5a1d1dcd1d2d0c273R68)

Brieuc-brd commented on addons/web_editor/static/src/scss/wysiwyg_snippets.scss:
```diff
@@ -1226,22 +1226,9 @@
-                }
+        .o_we_so_color_palette.o_we_user_value_widget:not(.d-none) {
+            + .o_we_user_value_widget:not(.o_we_so_color_palette) {
+                margin-left: $o-we-sidebar-content-field-multi-spacing * 4 !important;
```
We have made some tests with @mano-odoo and it seems to be fine.

Brieuc-brd commented on addons/website/views/website_templates.xml:
```diff
@@ -13,15 +13,15 @@
     </li>
     <li t-elif="show_dropdown" t-attf-class="#{item_class or ''} dropdown #{submenu.is_mega_menu and 'position-static'}">
-        <a t-attf-class="#{link_class or ''} dropdown-toggle #{submenu.is_mega_menu and 'o_mega_menu_toggle'} #{submenu._is_active() and 'active'}"
+        <a t-attf-class="#{link_class or ''} dropdown-toggle #{submenu.is_mega_menu and 'o_mega_menu_toggle'} #{submenu._is_active() and 'active'} #{dropdown_link_class}"
```
I suppose this could be the case in the future. I've renamed the classes.

Brieuc-brd commented on addons/website/views/website_templates.xml:
```diff
@@ -259,1105 +259,1190 @@
                 </t>
-                <ul class="nav navbar-nav navbar-expand d-none d-sm-block">
+                <!-- Extra elements -->
+                <ul class="navbar-nav align-items-center gap-2 flex-wrap justify-content-end ps-3">
```
I fixed this using `flex-shrink-0` but it creates an overflow when the viewport is close to the "medium" breakpoint.
I can't fix this with `flex-xx-wrap` otherwise the main navigation will expand and take up half the header.

This behavior could indicate to the user that there are too many elements in the navigation. But I'm not sure this is the right solution.

What do you think?
![Capture d’écran 2023-06-19 à 15 53 24](https://github.com/odoo/odoo/assets/80679690/3e35453f-241b-4844-a9d7-a03a23c4fc68)

Brieuc-brd commented:
Hello @stefanorigano thanks for your feedback 👍  I updated the PR (I just still have the 
ci/upgrade_enterprise 🔴  but I don't get how to make it green 🤔) :

> I think that the border option should affect the border-bottom property only

I agree, I tried to change it, but couldn't. I've put it in the list of technical issues in the task.

---

> The label should be place at the field's top

Fixed ✅ 

---

> The dropdown-menu "displacement" very problematic.
...
> 1. We "unfold " the menu by default (I suggest this)

If I understand correctly, if the user clicks on the button, it will fold the menu that was unfolded by default?
Isn't that weird? I tried the other solution using the full width dropup, what do you think?

---

> Searchbar's suggestions displace navigation links. Trying to hide them I clicked on big "X" closing the entire offcanvas menu instead .

I've forced the dropdown to have an absolute position, so there's no more weird expansion effect. What do you think?

---

>Not sure if already on master, but mega-menus are unfolded by default when inside the "+" dropdown

It's already in master, I'm not sure how to handle it. IMO since the `o_extra_menu_items` is handled in web I think it should be handled in another task, what do you think?

---

> For some reason "Contact us" becomes btn-lgon mobile.

Fixed ✅ 

---

> Stretch-menu template ( + Vertical-menu and others): "Over the content" -> Undesidered color changes

It seems to be a bug related to the header positioning feature, already present in master. I'll add it to the list of technical issues.

---

> User's dropdown-menu should be right-alighed

I've checked all the templates and corrected them where necessary.

---

>Search overlay modal: Border/overflow issue

Fixed ✅ 

---

>Social links (top-right). Aren't these supposed to become white?

Fixed ✅ 

---

pierre-pde commented:
@qsm-odoo , I think we now have something elligible for reviews :)
Maybe after 16.4 merge, up to you.
List of issues that will need rd-website help is available on task 3060986 under "Technical issues"

robinlej commented on addons/portal/views/portal_templates.xml:
```diff
@@ -51,21 +51,22 @@
-                        t-att-data-url_code="lg[1]">
-                            <span t-if="not no_text" t-esc="lg[2].split('/').pop()"/>
+                        t-att-data-url_code="lg[1]" t-att-title="lg[2].split('/').pop()">
+                            <span t-out="lg[2].split('/').pop()" t-attf-class="#{_txt_class}"/>
```
This is the line causing the bug of flag-only inline language selectors not working. :) You should keep `t-if="not no_text"`

qsm-odoo commented on addons/website/models/theme_models.py:
```diff
@@ -222,13 +222,6 @@ class Theme(models.AbstractModel):
         'website.template_header_hamburger',
         'website.template_header_vertical',
         'website.template_header_sidebar',
-        'website.template_header_slogan',
-        'website.template_header_contact',
-        'website.template_header_boxed',
-        'website.template_header_centered_logo',
-        'website.template_header_image',
-        'website.template_header_hamburger_full',
-        'website.template_header_magazine',
```
It's good that you split your work into multiple commits but ideally they should make sense on their own. In this first commit, you seem to remove template_header_default, but you keep it mentioned here in theme_models. You also remove it and its alignment options, making all the editor widgets empty (which does not make sense if I checkout this commit on its own)... but in a later commit you reintroduce it with the same xml id and with the same options for alignments. Could not you remove only what will be really removed in this commit and update what will be update into the related commit. This is important for review but also for the history. If I later stumble upon the "default header" commit, I cannot see what changed compared to the previous one... I need to find another commit for that :grimacing:

Brieuc-brd replied:
Hello @qsm-odoo ! 👋 

Indeed, I understand now this history issue 👍 
We've updated the commits structure and rebased the branch to fix this.
Let me know if this is ok for you :)

qsm-odoo reviewed (changes requested):
Hello @Brieuc-brd @robinlej 

Sorry but I am merging this first: https://github.com/odoo/odoo/pull/99732 (I just r+'ed). This will conflict with this work, could you anticipate and already rebase this branch on top of this other work ?

I also added a comment about the first commit (but the idea applies to the whole PR). If you can do something about it at the same time, that would be nice :pray: If not, I guess we'll do a merge commit in the end but it will make the history harder for everyone, same as my review :grimacing:

qsm-odoo commented:
@Brieuc-brd Thanks for the adaptation, I saw your message yesterday don't worry :wink: I'll have a look, just don't know when yet. If we do freeze tomorrow, I'll merge post-merge, like this week-end if nothing scares me too much :wink:

qsm-odoo commented on addons/website/tests/test_theme.py:
```diff
@@ -30,7 +30,7 @@ def _get_header_template_key():
                          "Only the default template should be active.")
 
-        key = 'website.template_header_magazine'
+        key = 'website.template_header_stretch'
```
This seems to be using a template that is not defined yet at this point in the github history + magazine does not exist anymore at this point either.

Brieuc-brd replied:
Good catch ! I changed it with the `vertical` template

qsm-odoo commented on addons/website/models/theme_models.py:
```diff
@@ -222,13 +222,6 @@ class Theme(models.AbstractModel):
         'website.template_header_hamburger',
         'website.template_header_vertical',
         'website.template_header_sidebar',
-        'website.template_header_slogan',
-        'website.template_header_contact',
-        'website.template_header_boxed',
-        'website.template_header_centered_logo',
-        'website.template_header_image',
```
I'll be testing migration at some point: what can I expect with this? Do users are reverted back to default header correctly if they choose one of those / customized stuff inside? :pray:

Brieuc-brd replied:
I think the best behavior would be to keep the old header selected by the user and if he replaces it with a new one, he wouldn't be able to retrieve the old one.
So does this mean we should keep the old templates? :/

qsm-odoo replied:
> So does this mean we should keep the old templates? :/

If we wanted to do that, kinda yes. We could potentially keep the forks users made but that would still mean having to keep the related CSS and the related mentions in the options XML. So we won't do that. Being reverted to the default one or ideally the closest new one is what should happen (thanks to the migration script). We can do it later.

qsm-odoo commented on addons/website/models/assets.py:
```diff
@@ -47,6 +47,7 @@ def make_scss_customization(self, url, values):
                 'header-sales_one-gradient': 'null',
                 'header-sales_two-gradient': 'null',
                 'header-sales_three-gradient': 'null',
+                'header-sales_four-gradient': 'null',
```
I guess it's ok but I am wondering why you needed multiple variables to handle a gradient of different templates. Could not you have shared them?

Brieuc-brd replied:
Indeed, it's better to use a shared gradient 👍 . I introduced `menu-secondary-gradient` in the `sales_one` commit and reused it in the three other `sales_*` templates.

qsm-odoo commented on addons/website/static/src/scss/website.scss:
```diff
@@ -48,7 +48,7 @@ $-seen-urls: ();
     // we have color variables which can contain a reference to a color
     // combination and that is the info we want in that case.
-    @each $key in ('menu', 'header-boxed', 'footer', 'copyright') {
+    @each $key in ('menu', 'header-sales_one', 'header-sales_two', 'header-sales_three', 'header-sales_four', 'footer', 'copyright') {
```
Not a fan but it's ok to me. I really don't like it in CSS classes though (I see you made some like that). But that's ok too, I will survive it :wink:

qsm-odoo commented on addons/website/views/snippets/snippets.xml:
```diff
@@ -954,6 +954,51 @@
+         data-no-check="true"
+         groups="website.group_website_designer">
+        <we-row string="Elements" class="o_we_full_row align-items-start">
+            <div class="w-100">
+                <div class="d-flex gap-1 mb-1">
```
I am pretty sure we don't need that it's suppose to be layed out automatically. Can't you just make multiple we-row? Would that not work? If not, it's alright, we'll review with OWL later anyway :wink:

Brieuc-brd replied:
I updated it using double `<we-row/>` :)

qsm-odoo commented on addons/website/static/src/js/editor/snippets.options.js:
```diff
@@ -2274,25 +2274,6 @@ options.registry.HeaderNavbar = options.Class.extend({
-            this.el.querySelector("[data-name='header_alignment_opt']").title =
-                _t("Mobile Alignment");
-        }
-    },
```
I think I'll end up doing a rebase-merge commit instead of a rebase-ff as I saw some mistakes like this one: here is another example of a commit which probably does not make sense on its own. Only removing code means one of your previous commit made that code useless and this is that commit which should contain this removal.

Brieuc-brd replied:
I got your point 👍 to fix that :

- I removed `"'magazine'", "'hamburger-full'", "'slogan'"` from the condition in https://github.com/odoo/odoo/pull/119650/commits/85c82c5df73da9263aefd07f55710bf376adb1ec
- Then I squashed the commit with https://github.com/odoo/odoo/pull/119650/commits/89c70544da40dce5420abd5744d77a915db30362 since they are related.

qsm-odoo commented on addons/website/views/website_templates.xml:
```diff
@@ -304,7 +304,7 @@
         <t t-set="_navbar_name" t-valuef="Mobile"/>
 
-        <div id="top_menu_container" class="container flex-wrap justify-content-between">
+        <div id="o_main_nav" class="container flex-wrap justify-content-between">
```
Renaming an ID is generally a bad idea for migration... but I guess that this work will generate many migration issues anyway so if there is a time to rename this ID, now is the time I guess. Let's keep it for now.

qsm-odoo reviewed (changes requested):
My courage meter is running low on gas so I did not take as much of a look as I wanted this week-end but scrolling through, nothing scares me too much. Well, migration aside: but I guess it's the same every year :grimacing: Thanks for the job :+1: 

I added some comments and questions. The history is better (thanks for that) but I think it contains mistake still, I'll probably do a rebase-merge. If you can take a look at my comments / adapt tomorrow, that's great. I will be deep-testing this tomorrow and we can hope to merge that on Tuesday :+1:

Brieuc-brd commented:
Hey @qsm-odoo, thanks for your review 💪 , I’ve updated the branch 👍 . 
Also, I haven’t rebase it to see the diff easily :
- https://github.com/odoo/odoo/compare/7f125612b51d6dd53263686129803341203fa057..0d98c27da291466b072a6759e91acd5678cd8082
- https://github.com/odoo/odoo/compare/0d98c27da291466b072a6759e91acd5678cd8082..cbe9a929bdd8304c875ed7a081b3d8c8effa0d1c (just forgot an empty line 😅 )

qsm-odoo reviewed (approved):
Ok, I am now confident enough to merge this without a merge commit. The history is still kinda weird but every commit works on its own even if the result at each step may be visually slightly broken.

Thank you for the work @Brieuc-brd @stefanorigano the new headers are great :+1: The code was clean from I saw (see below, I nearly changed nothing). It just took me hours to check and rewrite the git history. Maybe it's because I am a bit (a lot) maniac but what I always want to reach with a git history:
- A feature is not introduced by a commit and fixed by another one in the same work (dropdown position, mobile menu broken after removal in first commit, etc)
- A feature is not removed by a commit to be reintroduced by another one (removing 6 headers to reintroduce 6 new ones later)
- A feature does not introduce code that is unused by the commit itself (most first commits do that: create a mobile menu but start using it afterwards, create CSS classes but using it afterwards, etc)
- ... you get the idea.

The final state I reached still has every one of those problems, but reduced. It's ok. Like I said, maybe it's just me being maniac :wink:

Here is the history of my push-forces (cc @robinlej  Thank you for your work too :+1: for next reviews, try to check this too ->). Of course most push-forces are invisble diff since its about rewriting the history.

1. Just a rebase: https://github.com/odoo/odoo/compare/cbe9a929bdd8304c875ed7a081b3d8c8effa0d1c..5a6a8f8d3a99a8285b54ba78e7b4d46214f0e67d
2. Moving the template removal at the end, no diff: https://github.com/odoo/odoo/compare/5a6a8f8d3a99a8285b54ba78e7b4d46214f0e67d..de14ec8f4326539c10661c62f5bd6ebd40d0424d
3. Squashing the new mobile menu feature with the removal of the old one: https://github.com/odoo/odoo/compare/de14ec8f4326539c10661c62f5bd6ebd40d0424d..d2cf82134e8562ee10a456fd6e6d0c641162de9d
4. Fixing small issues with headers removal and trying to couple the old headers to equivalent new ones: https://github.com/odoo/odoo/compare/d2cf82134e8562ee10a456fd6e6d0c641162de9d..f52988e5f585d92d407d72adbc7538eecb162cd6
    - I chose "sales_2" to be the old "slogan", I kept your new xml_id (adapted upgrade script) but remove the old one in the same commit as the new one and made it so someone with "slogan" will have "sales_2" post update.
    - Same as above for "contact" -> "sales_3"
    - Same as above for "centered_logo" -> "search"
    - I restored the same xml_id "header_templated_boxed" and css variable for the new "rounded box" one.
    - image -> still no replacement
    - hamburger-full -> still no replacement
    - template_header_magazine -> still no replacement
5. Reviewing each commit one by one (small changes) and fixing the errors I made myself in previous re-organisation: https://github.com/odoo/odoo/compare/f52988e5f585d92d407d72adbc7538eecb162cd6..19b300919b51196a777befcee40edd2fa8a9d899
6. Adding new commit of mine:  https://github.com/odoo/odoo/compare/19b300919b51196a777befcee40edd2fa8a9d899..bdbfc3043274c365c9b360d5a28391f59189a268
7. Removing the last commit of @robinlej about searchbar options (sorry it took so long to check @robinlej). I checked for like half an hour but I don't have a solution either for now. But I also don't think it's ok to duplicate all options into toggleable views... to think about. At worst, we can have no option for 17.0 I think:  https://github.com/odoo/odoo/compare/bdbfc3043274c365c9b360d5a28391f59189a268..39616e46f13022699f8c221dcd80e71b7f98c22d
8. Added work from @SergeBayet extracted from https://github.com/odoo/odoo/pull/132023: https://github.com/odoo/odoo/compare/39616e46f13022699f8c221dcd80e71b7f98c22d..f7c16ce1f4c761aa330db2b0cb195d818629589e - Thanks for your work too, it was good :+1: Some things though:
    - You forgot to remove language_selector_add_language after no template used it anymore
    - Could you check the footer language selection option? It's visually misindented.
    - In the end, I chose to keep the possibility of inline language selector in the header. With the new header templates, it works great, especially when inside the offcanvas menu IMO.
9. Final rebase: https://github.com/odoo/odoo/compare/f7c16ce1f4c761aa330db2b0cb195d818629589e..fab08341aa67127307d2e1ced49c8b17c799355c

There are of course things to improve, things I missed and bugs I potentially created myself doing this. If you see an error, tell me ! In any case, I already wrote those things to consider:
- `include o-bg-color($-menu-color);` -> should not it be `apply-colors` ?
- div in ul for language selector in 16.4? Should it be fixed somehow?
- fa-stack in first commit is breaking the layout. Using it in the snippet options of social media may not be good and m-1 should be removable in any case.
- Edition of call to action buttons may be broken. I need to test more but trying it, sometimes it saved editor preview (cyan colors) inside the button, etc :grimacing: 
- Check if header text can be edited properly. Seeing the DOM I am not sure it will always work great. There should be an wrapper which is the editable element.
- To check: should we/do we already reset all views related to option when disabling/enabling them. It easy IMO to break a social media and want to restore the default view.
- Some translation in header text are not done properly I think
- Should not we share sales_xx templates colors (like you did for gradients @Brieuc-brd)
- $-hamburger-left/center/right seems to be dead code mostly?
- Sidebar on the right was lost as a feature... wanted? I don't mind because it's really not an important header to me but, still.
- I understand removing hamburger-full, magazine and for the 3 other ones I found an ok-replacement but for image, I feel it's a bit sad to have lost it. Any reason?
- Fix dropdown when menu dark -> backport?
- "Content saved" message annoying + "loading effect" annoying when playing with header options, to improve on our side.
- Default header, social media + cart + wishlist -> bad spacing IMO -> we probably should review all spacings around header templates to be sure.
- Not sure what the fix commit about search suggestions forced in absolute was about but... I kept it as you did :wink: 
- Need to check for t-cache issues to be sure
- Create a submenu, check the "mobile" menu in desktop mode (hamburger), centered option -> broken
- o_header_force_no_radius: would have been nice not to do that with a class IMO but rather a list edit mode side only (imp for later with many things like that outside of this work)
- Enable over the content by default works only for current page of course... could be strange?
- Create css alias for 3 replaced headers somehow, the mapping I did works: someone with "slogan" enabled will have "sales_2" views after the update but not the right CSS variable

Anyways :smile: If anyone want to go through that list, say it first ! :smile: Once all our features reach master, I'll ask the team to hunt down bugs anyway :)

Again, thanks everyone, this looks good :+1: 

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
